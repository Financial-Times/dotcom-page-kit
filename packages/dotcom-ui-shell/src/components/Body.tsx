import React from 'react'

export type TBodyProps = {
  contents?: string | React.ReactNode
}

const styles = {
  display: 'contents'
}

function Body({ contents }: TBodyProps) {
  if (typeof contents === 'string') {
    return <div style={styles} dangerouslySetInnerHTML={{ __html: contents }} />
  } else {
    return <div style={styles}>{contents}</div>
  }
}

export default Body
