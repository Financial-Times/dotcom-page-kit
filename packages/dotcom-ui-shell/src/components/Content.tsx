import React from 'react'

export type TContentProps = {
  contents?: string | React.ReactNode
}

const styles = {
  display: 'contents'
}

function Content({ contents }: TContentProps) {
  if (typeof contents === 'string') {
    return <div style={styles} dangerouslySetInnerHTML={{ __html: contents }} />
  }

  if (React.isValidElement(contents)) {
    return contents
  }

  return null
}

export default Content
