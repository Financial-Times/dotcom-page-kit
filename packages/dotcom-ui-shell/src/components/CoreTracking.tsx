import React from 'react'

const pixel = 'https://spoor-api.ft.com/px.gif?data='

export default function CoreTracking({ context }) {
  const trackingData = {
    category: 'page',
    action: 'view',
    system: {
      source: 'non-ctm'
    },
    context: {
      ...context,
      product: 'next',
      data: { source: 'SOURCE' }
    }
  }

  const encodedTrackingData = encodeURIComponent(JSON.stringify(trackingData))

  // NOTE: This function will be stringified and embedded so use ES5 only!
  function coreExperience() {
    if (/\bcore\b/.test(document.documentElement.className)) {
      var currentScript = document.scripts[document.scripts.length - 1]
      var img = new Image()

      img.src = currentScript.getAttribute('data-pixel-src')
    }
  }

  return (
    <React.Fragment>
      <script
        data-pixel-src={pixel + encodedTrackingData.replace('SOURCE', 'core-experience')}
        dangerouslySetInnerHTML={{ __html: '(' + coreExperience.toString() + ')();' }}
      />
      <noscript>
        <img src={pixel + encodedTrackingData.replace('SOURCE', 'no-js')} />
      </noscript>
    </React.Fragment>
  )
}
