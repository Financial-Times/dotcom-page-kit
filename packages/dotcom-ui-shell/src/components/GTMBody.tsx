import React from 'react'
import { TFlagsData } from '@financial-times/dotcom-ui-flags/src/types'

// This component is maintained by the ads team
const GTMBody = ({ flags }: { flags: TFlagsData }) => {
  if (!flags.enableGTM) {
    return null
  }

  return (
    <noscript>
      <iframe
        src="https://www.googletagmanager.com/ns.html?id=GTM-NWQJW68"
        height="0"
        width="0"
        style={{
          display: 'none',
          visibility: 'hidden'
        }}
      />
    </noscript>
  )
}

GTMBody.defaultProps = {
  flags: {}
}

export default GTMBody
