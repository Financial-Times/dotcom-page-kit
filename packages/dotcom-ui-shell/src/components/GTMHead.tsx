import React from 'react'
import { TFlagsData } from '@financial-times/dotcom-ui-flags/src/types'

// This component is maintained by the ads team
type Props = {
  flags: TFlagsData
  appName?: string
}

// This component is maintained by the ads team
const GTMHead = ({ flags, appName }: Props) => {
  if (!flags.enableGTM) return null

  // 1) If appName is artcile-partner-content use the PC GTM instance
  const PCAppOverride =
    appName === 'article-partner-content' ? 'https://www.googletagmanager.com/gtm.js?id=GTM-12345' : null

  // 2) If appName is not Partner Content, use the standard FT.com GTM isntance and use the first-party GTM domain if flag is set.
  const src =
    PCAppOverride ??
    (flags['ads-first-party-gtm']
      ? 'https://www.ft.com/page-resources/'
      : 'https://www.googletagmanager.com/gtm.js?id=GTM-NWQJW68')

  const tagManager = `(function(w,d,s,l){
    w[l]=w[l]||[];w[l].push({'gtm.start': new Date().getTime(),event:'gtm.js'});
    var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';
    j.async=true;
    j.src='${src}'+dl;
    f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer');`

  return <script dangerouslySetInnerHTML={{ __html: tagManager }} />
}

GTMHead.defaultProps = {
  flags: {}
}

export default GTMHead
