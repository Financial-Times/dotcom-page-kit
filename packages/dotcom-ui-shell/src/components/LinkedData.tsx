import React from 'react'

export type TLinkedDataProps = {
  jsonLd?: { [key: string]: any }
}

const LinkedData = ({ jsonLd }: TLinkedDataProps) => (
  <React.Fragment>
    {Array.isArray(jsonLd) &&
      jsonLd.map((data, i) => (
        <script
          key={`jsonld-${i}`}
          type="application/ld+json"
          dangerouslySetInnerHTML={{
            __html: JSON.stringify(data)
          }}
        />
      ))}
    <script
      type="application/ld+json"
      dangerouslySetInnerHTML={{
        __html: JSON.stringify({
          '@context': 'http://schema.org',
          '@type': 'WebSite',
          name: 'Financial Times',
          alternateName: 'FT.com',
          url: 'http://www.ft.com'
        })
      }}
    />
  </React.Fragment>
)

LinkedData.defaultProps = {
  jsonLd: []
}

export default LinkedData
