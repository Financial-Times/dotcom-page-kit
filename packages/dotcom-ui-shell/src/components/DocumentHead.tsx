import React from 'react'
import imageServiceIconURL from '../lib/imageServiceIconURL'
import OpenGraph, { TOpenGraphProps } from './OpenGraph'
import LinkedData, { TLinkedDataProps } from './LinkedData'
import GTMHead from './GTMHead'

export type TDocumentHeadProps = TOpenGraphProps &
  TLinkedDataProps & {
    flags?
    description?: string
    facebookPage?: string
    googleSiteVerification?: string
    metaTags?: Array<{ [key: string]: any }>
    pageTitle: string
    robots?: string
    siteTitle?: string
    twitterSite?: string
    canonicalURL?: string
    additionalMetadata?: React.ReactNode
  }

const DocumentHead = (props: TDocumentHeadProps) => (
  <React.Fragment>
    <meta charSet="utf-8" />
    <meta httpEquiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>{props.pageTitle ? `${props.pageTitle} | ${props.siteTitle}` : props.siteTitle}</title>

    {props.description && <meta name="description" content={props.description} />}

    {props.canonicalURL && <link rel="canonical" href={props.canonicalURL} />}

    {/* SEO */}
    <meta name="robots" content={props.robots} />
    <meta name="google-site-verification" content={props.googleSiteVerification} />
    {props.metaTags.map((attributes, i) => (
      <meta key={`meta-${i}`} {...attributes} />
    ))}
    <LinkedData jsonLd={props.jsonLd} />

    {/* social media */}
    <meta property="fb:pages" content={props.facebookPage} />
    <meta property="twitter:site" content={props.twitterSite} />
    <OpenGraph openGraph={props.openGraph} />

    {/* packaging */}
    <link
      rel="icon"
      type="image/png"
      href={imageServiceIconURL('ftlogo-v1:brand-ft-logo-square-coloured', 32)}
      sizes="32x32"
    />
    <link
      rel="icon"
      type="image/png"
      href={imageServiceIconURL('ftlogo-v1:brand-ft-logo-square-coloured', 194)}
      sizes="194x194"
    />
    <link
      rel="apple-touch-icon"
      href={imageServiceIconURL('ftlogo-v1:brand-ft-logo-square-coloured', 180)}
      sizes="180x180"
    />

    {/* We can't add an option for every single metadata option so allow custom elements to be inserted*/}
    {props.additionalMetadata}
    <GTMHead flags={props.flags} />
  </React.Fragment>
)

DocumentHead.defaultProps = {
  description:
    'News, analysis and comment from the Financial Times, the world ºs leading global business publication',
  facebookPage: '8860325749',
  googleSiteVerification: '4-t8sFaPvpO5FH_Gnw1dkM28CQepjzo8UjjAkdDflTw',
  metaTags: [],
  jsonLd: [],
  robots: 'index,follow',
  siteTitle: 'Financial Times',
  twitterSite: '@FinancialTimes',
  additionalMetadata: null,
  flags: {}
}

export default DocumentHead
