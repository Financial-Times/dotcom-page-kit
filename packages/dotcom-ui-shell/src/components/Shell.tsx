import React from 'react'
import Body, { TBodyProps } from './Body'
import DocumentHead, { TDocumentHeadProps } from './DocumentHead'
import StyleSheets, { TStylesheetProps } from './StyleSheets'
import { AppContextEmbed, TAppContextProps } from '@financial-times/dotcom-ui-app-context'
import { FlagsEmbed, TFlagsEmbedProps } from '@financial-times/dotcom-ui-flags'
import { Bootstrap, TBootstrapProps } from '@financial-times/dotcom-ui-bootstrap'
import * as PolyfillService from '@financial-times/dotcom-ui-polyfill-service'
import formatAttributeNames, { TAttributeData } from '../lib/formatAttributeNames'
import CoreTracking from './CoreTracking'

type TShellProps = TDocumentHeadProps &
  TAppContextProps &
  TStylesheetProps &
  TBodyProps &
  TFlagsEmbedProps &
  TBootstrapProps & {
    children?: any
    initialProps?: any
    bodyAttributes?: TAttributeData
    htmlAttributes?: TAttributeData
    appPolyfillsCore?: Array<String>
    appPolyfillsEnhanced?: Array<String>
  }

function Shell(props: TShellProps) {
  const corePolyfills = [PolyfillService.core(props.appPolyfillsCore)]
  const enhancedPolyfills = [PolyfillService.enhanced(props.appPolyfillsEnhanced)]

  const coreScripts = [...corePolyfills, ...props.coreScripts]
  const enhancedScripts = [...enhancedPolyfills, ...props.enhancedScripts]

  return (
    <html
      {...formatAttributeNames(props.htmlAttributes)}
      lang="en-GB"
      className="no-js core"
      style={{
        // TODO: abstract styles into 'core branding' package
        // Enable use of 100vw which does not account for the scroll bar
        overflowX: 'hidden',
        backgroundColor: '#fff1e5',
        color: '#33302e'
      }}>
      <head>
        <DocumentHead {...props} />
        {/* TODO: refactor initial props, flags data, and context data to the bottom of body */}
        <script
          id="initial-props"
          type="application/json"
          dangerouslySetInnerHTML={{ __html: JSON.stringify(props.initialProps) }}
        />
        <FlagsEmbed flags={props.flags} />
        <AppContextEmbed context={props.context} />
        <StyleSheets stylesheets={props.stylesheets} criticalStyles={props.criticalStyles} />
        <Bootstrap coreScripts={coreScripts} enhancedScripts={enhancedScripts} trackErrors={true} />
        <CoreTracking context={props.context} />
      </head>
      <Body {...formatAttributeNames(props.bodyAttributes)} contents={props.contents || props.children} />
    </html>
  )
}

Shell.defaultProps = {
  coreScripts: [],
  enhancedScripts: [],
  htmlAttributes: {},
  bodyAttributes: {}
}

export { Shell, TShellProps }
