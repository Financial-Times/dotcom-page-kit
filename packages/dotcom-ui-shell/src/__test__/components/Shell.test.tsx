import React from 'react'
import renderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import { Shell } from '../../components/Shell'

describe('dotcom-ui-shell/src/components/Shell', () => {
  it('should define all props as optional except the `pageTitle` prop', () => {
    new ShallowRenderer().render(<Shell pageTitle="Foo" />)
  })

  it('renders boolean boolean html data attributes correctly', () => {
    const htmlAttributes = { dataIsProduction: '', dataIsDevelopment: false }
    const htmlTag = renderer.create(<Shell pageTitle="Foo" htmlAttributes={htmlAttributes} />).toJSON()

    // where react is concerned, a `true` boolean data attribute
    // is one where the attribute value is an empty string (because
    // it is not possible to render the attribute without a value),
    // and a `false` boolean data attribute is one where the attribute
    // has not been specified altogether
    expect(htmlTag.props).toHaveProperty('data-is-production', '')
    expect(htmlTag.props).not.toHaveProperty('data-is-development')
  })

  it('renders the GTM script when the enableGTM flag is on', () => {
    const tree = renderer.create(<Shell pageTitle="Foo" flags={{ enableGTM: true }} />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
