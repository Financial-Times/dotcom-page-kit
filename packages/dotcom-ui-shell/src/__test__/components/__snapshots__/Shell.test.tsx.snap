// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dotcom-ui-shell/src/components/Shell renders the GTM script when the enableGTM flag is on 1`] = `
<html
  className="no-js core o-typography--loading-sans o-typography--loading-sans-bold o-typography--loading-display o-typography--loading-display-bold"
  data-o-component="o-typography"
  lang="en-GB"
  style={
    Object {
      "backgroundColor": "#fff1e5",
      "color": "#33302e",
      "overflowX": "hidden",
    }
  }
>
  <head>
    <meta
      charSet="utf-8"
    />
    <meta
      content="IE=edge"
      httpEquiv="X-UA-Compatible"
    />
    <meta
      content="width=device-width, initial-scale=1"
      name="viewport"
    />
    <title>
      Foo
    </title>
    <meta
      content="index,follow,max-snippet:200,max-image-preview:large"
      name="robots"
    />
    <meta
      content="4-t8sFaPvpO5FH_Gnw1dkM28CQepjzo8UjjAkdDflTw"
      name="google-site-verification"
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{\\"@context\\":\\"http://schema.org\\",\\"@type\\":\\"WebSite\\",\\"name\\":\\"Financial Times\\",\\"alternateName\\":\\"FT.com\\",\\"url\\":\\"http://www.ft.com\\"}",
        }
      }
      type="application/ld+json"
    />
    <meta
      content="8860325749"
      property="fb:pages"
    />
    <meta
      content="@FinancialTimes"
      property="twitter:site"
    />
    <meta
      content="app-id=1200842933"
      name="apple-itunes-app"
    />
    <link
      href="https://www.ft.com/__origami/service/image/v2/images/raw/ftlogo-v1%3Abrand-ft-logo-square-coloured?source=update-logos&format=svg"
      rel="icon"
      type="image/svg+xml"
    />
    <link
      href="https://www.ft.com/__origami/service/image/v2/images/raw/ftlogo-v1%3Abrand-ft-logo-square-coloured?source=update-logos&format=png&width=32&height=32"
      rel="alternate icon"
      sizes="32x32"
      type="image/png"
    />
    <link
      href="https://www.ft.com/__origami/service/image/v2/images/raw/ftlogo-v1%3Abrand-ft-logo-square-coloured?source=update-logos&format=png&width=194&height=194"
      rel="alternate icon"
      sizes="194x194"
      type="image/png"
    />
    <link
      href="https://www.ft.com/__origami/service/image/v2/images/raw/ftlogo-v1%3Abrand-ft-logo-square-coloured?source=update-logos&format=png&width=180&height=180"
      rel="apple-touch-icon"
      sizes="180x180"
    />
    <link
      href="https://www.ft.com/__assets/creatives/manifest/manifest-v6.json"
      rel="manifest"
    />
    <link
      href="https://spoor-api.ft.com"
      rel="preconnect"
    />
    <link
      crossOrigin="use-credentials"
      href="https://session-next.ft.com"
      rel="preconnect"
    />
    <link
      href="https://ads-api.ft.com"
      rel="preconnect"
    />
    <link
      href="https://securepubads.g.doubleclick.net"
      rel="preconnect"
    />
    <link
      as="font"
      crossOrigin="anonymous"
      href="https://www.ft.com/__origami/service/build/v3/font?font_format=woff2&font_name=MetricWeb-Regular&system_code=page-kit&version=1.12"
      rel="preload"
      type="font/woff2"
    />
    <link
      as="font"
      crossOrigin="anonymous"
      href="https://www.ft.com/__origami/service/build/v3/font?font_format=woff2&font_name=MetricWeb-Semibold&system_code=page-kit&version=1.12"
      rel="preload"
      type="font/woff2"
    />
    <link
      as="font"
      crossOrigin="anonymous"
      href="https://www.ft.com/__origami/service/build/v3/font?font_format=woff2&font_name=FinancierDisplayWeb-Regular&system_code=page-kit&version=1.12"
      rel="preload"
      type="font/woff2"
    />
    <link
      as="font"
      crossOrigin="anonymous"
      href="https://www.ft.com/__origami/service/build/v3/font?font_format=woff2&font_name=FinancierDisplayWeb-Bold&system_code=page-kit&version=1.12"
      rel="preload"
      type="font/woff2"
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
      id="initial-props"
      type="application/json"
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{\\"trackErrors\\":true,\\"core\\":[],\\"enhanced\\":[]}",
        }
      }
      id="page-kit-bootstrap-config"
      type="application/json"
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": ";(function () {
  var doc = document.documentElement
  var isEnhanced = isEnhancedBrowser()
  var scriptsConfig = getScriptsConfig()
  var scriptsToLoad = []
  var currentScript = document.scripts[document.scripts.length - 1]

  doc.className = doc.className.replace('no-js', 'js')

  if (isEnhanced) {
    doc.className = doc.className.replace('core', 'enhanced')
    Array.prototype.push.apply(scriptsToLoad, scriptsConfig.enhanced)
  } else {
    Array.prototype.push.apply(scriptsToLoad, scriptsConfig.core)
  }

  for (var i = 0, len = scriptsToLoad.length; i < len; i++) {
    loadScript(scriptsToLoad[i])
  }

  function scriptLoadError(error) {
    var script = error.target ? error.target.src : null

    if (script) {
      console.error('The script ' + script + ' failed to load') // eslint-disable-line no-console
    }

    if (/enhanced/.test(doc.className)) {
      console.warn('Script loading failed, reverting to core experience') // eslint-disable-line no-console
      doc.className = doc.className.replace('enhanced', 'core')
    }

    if (scriptsConfig.trackErrors) {
      addErrorTrackingPixel(script)
    }
  }

  function loadScript(src) {
    var script = document.createElement('script')
    script.onerror = scriptLoadError
    script.async = false
    script.src = src
    currentScript.parentNode.insertBefore(script, currentScript)
  }

  // \\"Cut the mustard\\" test
  // by Maggie Allen and Matt Hinchliffe November 2018
  function isEnhancedBrowser() {
    var script = document.createElement('script')
    var input = document.createElement('input')

    return (
      'visibilityState' in document && // not supported by old Android (4.0-4.4) without a prefix
      'indeterminate' in input && // not supported by BB 10
      'flex' in doc.style && // not supported by old Safari (< 9) or IE 6-10
      'async' in script // not supported by old Opera (Presto engine < 15)
    )
  }

  function getScriptsConfig() {
    var scriptsConfigEl = document.getElementById('page-kit-bootstrap-config')
    var scriptsConfig = { core: [], enhanced: [], trackErrors: false }

    if (scriptsConfigEl) {
      try {
        scriptsConfig = JSON.parse(scriptsConfigEl.innerHTML)
      } catch (error) {
        console.error('Bootstrap configuration error', error) // eslint-disable-line no-console
      }
    }

    return scriptsConfig
  }

  function addErrorTrackingPixel(script) {
    var img = new Image()

    var data = JSON.stringify({
      category: 'javascript',
      action: 'load-error',
      system: {
        source: 'page-kit'
      },
      context: {
        script: script
      }
    })

    img.src = 'https://spoor-api.ft.com/px.gif?data=' + encodeURIComponent(data)
  }
})()
",
        }
      }
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "(function(w,d,s,l){
    w[l]=w[l]||[];w[l].push({'gtm.start': new Date().getTime(),event:'gtm.js'});
    var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';
    j.async=true;
    j.src='https://www.googletagmanager.com/gtm.js?id=GTM-NWQJW68'+dl;
    f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer');",
        }
      }
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "(function loadCustomFonts() {
  var rootElement = document.documentElement;
  if (/(^|\\\\s)o-typography-fonts-loaded=1(;|$)/.test(document.cookie)) {
    var fontLabels = ['sans', 'sans-bold', 'display', 'display-bold'];
    for (var i = 0; i < fontLabels.length; i++) {
      rootElement.className = rootElement.className.replace('o-typography--loading-' + fontLabels[i], '');
    }
  }
}());",
        }
      }
    />
  </head>
  <body>
    <noscript>
      <iframe
        height="0"
        src="https://www.googletagmanager.com/ns.html?id=GTM-NWQJW68"
        style={
          Object {
            "display": "none",
            "visibility": "hidden",
          }
        }
        width="0"
      />
    </noscript>
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
      id="page-kit-app-context"
      type="application/json"
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{\\"enableGTM\\":true}",
        }
      }
      id="page-kit-flags-embed"
      type="application/json"
    />
  </body>
</html>
`;
